### PR TITLE
Filter out duplicated defined custom-entries and only take the custom one

### PR DESCRIPTION
**Pull request by ymo-sci**
### Description
This is needed to "override"  base defined entries in a custom file and not duplicate it in the system (where the difference is not easily visible)

### Checklist

- [x] Code change has been tested and works locally
- [x] Code was formatted via IntelliJ and follows SonarLint & [best practices](https://scireum.myjetbrains.com/youtrack/articles/MISC-A-16/CodeStyle-JavaDoc)
- [ ] Patch Tasks: Is local execution of Patch Tasks necessary? If so, please also mark the PR with the tag.

### CODE REVIEW

File: src/main/java/sirius/biz/codelists/CustomLookupTable.java [hunk 1]
```diff
@@ -116,6 +130,13 @@ public int count() {
...
     protected Stream<LookupTableEntry> performQuery(String language, String lookupPath, String lookupValue) {
         return Stream.concat(customTable.performQuery(language, lookupPath, lookupValue),
-                             baseTable.performQuery(language, lookupPath, lookupValue));
+                             baseTable.performQuery(language, lookupPath, lookupValue))
+                     .collect(filterCustomDuplicateCollector())
+                     .values()
+                     .stream();
+    }
+
+    private Collector<LookupTableEntry, ?, Map<String, LookupTableEntry>> filterCustomDuplicateCollector() {
```
Comment (idlira) [inline src/main/java/sirius/biz/codelists/CustomLookupTable.java:139]: I would expect a method called filter*** to be used with a Stream.filter.

File: src/main/java/sirius/biz/codelists/CustomLookupTable.java [hunk 2]
```diff
@@ -90,22 +95,31 @@ protected Stream<LookupTableEntry> performSuggest(Limit limit, String searchTerm
         return Stream.concat(customTable.performSuggest(Limit.UNLIMITED, searchTerm, language),
                              baseTable.performSuggest(Limit.UNLIMITED, searchTerm, language))
                      .skip(limit.getItemsToSkip())
-                     .limit(limit.getMaxItems() == 0 ? Long.MAX_VALUE : limit.getMaxItems());
+                     .limit(limit.getMaxItems() == 0 ? Long.MAX_VALUE : limit.getMaxItems())
+                     .collect(filterCustomDuplicateCollector())
+                     .values()
```
Comment (idlira) [inline src/main/java/sirius/biz/codelists/CustomLookupTable.java:100]: Why not `.filter` with a Predicate? Fells not performant to collect everything and build a new stream.

**Review by idlira — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/codelists/CustomLookupTable.java (hunk 2 above)
Comment (jmuscireum) [reply src/main/java/sirius/biz/codelists/CustomLookupTable.java:100]: filter can't access the other elements in the stream. Another solution would be something like:

```java
Map<String, LookupTableEntry> results = new LinkedHashMap<>();

customTable.performSuggest(Limit.UNLIMITED, searchTerm, language)
           .forEach(entry -> results.put(entry.getCode(), entry));

baseTable.performSuggest(Limit.UNLIMITED, searchTerm, language)
         .forEach(entry -> results.putIfAbsent(entry.getCode(), entry));

return results.values().stream().limit(limit.getMaxItems() == 0 ? Long.MAX_VALUE : limit.getMaxItems());
```

Implementing `LookupTableEntry#equals` which only compares the code and then using distinct here might be another solution.

In any case the limit needs to be the last part in the chain.

Comment (idlira) [reply src/main/java/sirius/biz/codelists/CustomLookupTable.java:99]: I think you can try with a `.reduce()` As soon as you get 2 items with the same code, you gotta decide which one to keep.

Comment (idlira) [reply src/main/java/sirius/biz/codelists/CustomLookupTable.java:99]: or maybe something like that could work as well. So you only need the codes in a temp set.

```java
Set<String> codes = new HashSet<>();
        return Stream.concat(customTable.performSuggest(Limit.UNLIMITED, searchTerm, language),
                             baseTable.performSuggest(Limit.UNLIMITED, searchTerm, language))
                     .skip(limit.getItemsToSkip())
                     .filter(codes::add)
                     .limit(limit.getMaxItems() == 0 ? Long.MAX_VALUE : limit.getMaxItems());
```

**Review by idlira — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by jmuscireum — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Pull request merged**